spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -247,37 +295,66 @@ case class DeltaSharingSource(
       isStartingVersion: Boolean,
       endingVersionForQuery: Long): Unit = {
     lastQueryTableTimestamp = System.currentTimeMillis()
+    minUrlExpirationTimestamp = None
     if (isStartingVersion) {
       // If isStartingVersion is true, it means to fetch the snapshot at the fromVersion, which may
       // include table changes from previous versions.
       val tableFiles = deltaLog.client.getFiles(
         deltaLog.table, Nil, None, Some(fromVersion), None, None
       )
       latestRefreshFunc = () => {
-        deltaLog.client.getFiles(
+        lastQueryTableTimestamp = System.currentTimeMillis()
+        val files = deltaLog.client.getFiles(
           deltaLog.table, Nil, None, Some(fromVersion), None, None
-        ).files.map { f =>
+        ).files
+        var minUrlExpiration: Option[Long] = None
+        val idToUrl = files.map { f =>
+          if (f.expirationTimestamp != null) {
+            minUrlExpiration = if (minUrlExpiration.isDefined &&
+              minUrlExpiration.get < f.expirationTimestamp) {
+              minUrlExpiration
+            } else {
+              Some(f.expirationTimestamp)
+            }
+          }
           f.id -> f.url
         }.toMap
+
+        refreshSortedFetchedFiles(idToUrl)
+        minUrlExpirationTimestamp = minUrlExpiration
```

Comment:
Either make this atomic, or protect it with synchronized.

Reply:
added synchronized in all places.